### DISCUSSION TITLE
[Feature Request] scaling factor for initializers

### DISCUSSION THREAD

**Discussion by PhilipVinc** [google/flax#2714] [Q&A]
Machine-Learning practitioners are aware that initialisation functions must be properly rescaled depending on the variance of the output of the nonlinearity.

Howvever, scientistis like Mathematicians/Biologists/Physicists and many others who start looking into _Differentiable programming_ and/or _Scientific Machine Learning_ (or, as I call it, let's throw Neural networks at our problems and see how they perform) are relatively 'illiterate' about this kind of things. 
If `Flax` is supposed to be targeted at this kind of users, maybe you could do something to more prominently make users aware of this issue?  
I've seen many (NetKet) users mindlessly changing activation functions, and they had no idea they should also have rescaled 
Maybe mentioning this in a tutorial/notebook might already be usefull.

Also, I recently was made aware of [`torch.nn.init.calculate_gain(nonlinearity, param=None)`](https://pytorch.org/docs/stable/nn.init.html) which is an useful utility returning the variance of the output of the activation fucntions. 
Maybe flax could also export something similar?

**Comment by PhilipVinc** [comment]
Maybe, a `DenseActivation` layer which automatically rescaled the initialiser depending on the nonlinearity, if it is known?
That wouldn't work all the time, and maybe is too smart, but it would be very beginner friendly.

**Comment by jheek** [comment]
>> Machine-Learning practitioners are aware that initialisation functions must be properly rescaled depending on the variance of the output of the nonlinearity.

I don't think this is still an up-to-date statement. There was a time when network training was very sensitive to the initializers because there were no skip connections or normalization.
Nowadays both of these techniques are almost universally applied and have made tuning the initializers largely irrelevant. It's been quite a while since I have seen a network that *did* have proper initialization correcting for the activation function.

That said, computing the gain might still be interesting, but it isn't essential like it used to be.

**Comment by marcvanzee** [comment]
@PhilipVinc I transfered your issue to a discussion, since it seems useful for users to be aware of this, but we probably will not implement any utility functions / custom classes for rescaling initializers soon. However, if this comes up more often we might reconsider, thanks for raising it!
